### PR TITLE
add handler factory interface and abstract implementation

### PR DESCRIPTION
**Pull request by aptasinskij**

### CODE REVIEW

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 1]
```diff
@@ -6,15 +6,15 @@
...
+public abstract class AbstractRequestHandler<T extends RequestData, R extends ResponseData> implements RequestHandler {
 
 	@Override
 	public Response<R> handle(Request<?> msg) {
-		return null;
+	    Request<T> request = defineRequestDataType(msg);
+		return continueHandling(request);
 	}
 
+	abstract Request<T> defineRequestDataType(Request<?> msg);
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij This should be done here. Not in child classes.

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 2]
```diff
@@ -1,20 +1,38 @@
...
+    }
+
 	@Override
 	public Response<R> handle(Request<?> msg) {
 	    Request<T> request = defineRequestDataType(msg);
 		return continueHandling(request);
 	}
 
-	abstract Request<T> defineRequestDataType(Request<?> msg);
+	Request<T> defineRequestDataType(Request<?> msg){
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij why default scope ?

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 3]
```diff
@@ -1,20 +1,38 @@
...
+        JacksonDataMapper mapper = new JacksonDataMapper();
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij Why new mapper? We already have module that provides mapper.

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 4]
```diff
@@ -1,35 +1,39 @@
...
 	public Response<R> handle(Request<?> msg) {
 	    Request<T> request = defineRequestDataType(msg);
-		return continueHandling(request);
+		return handleRequest(request);
 	}
 
-	Request<T> defineRequestDataType(Request<?> msg){
-        JacksonDataMapper mapper = new JacksonDataMapper();
+	private Request<T> defineRequestDataType(Request<?> msg){
         T data = mapper.mapData(mapper.objectToString(msg.getData()), type);
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij why two conversions? We need only one.

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 5]
```diff
@@ -1,35 +1,39 @@
...
 import com.softgroup.common.protocol.Response;
 import com.softgroup.common.protocol.ResponseData;
+import org.springframework.beans.factory.annotation.Autowired;
 
 public abstract class AbstractRequestHandler<T extends RequestData, R extends ResponseData> implements RequestHandler {
 
     private final Class<T> type;
 
+    @Autowired
+    DataMapper mapper;
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij Why default scope ?

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 6]
```diff
@@ -1,20 +1,39 @@
...
-		return null;
-	}
+public abstract class AbstractRequestHandler<T extends RequestData, R extends ResponseData> implements RequestHandler {
+
+    private final Class<T> type;
+
+    @Autowired
+    private DataMapper mapper;
+
+    public AbstractRequestHandler(Class<T> type) {
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij if you use constructor - then you should use constructor injection as well.

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 7]
```diff
@@ -1,20 +1,39 @@
...
+    }
 
 	@Override
 	public Response<R> handle(Request<?> msg) {
-		return null;
+	    Request<T> request = defineRequestDataType(msg);
+		return handleRequest(request);
 	}
 
+	private Request<T> defineRequestDataType(Request<?> msg){
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij you are not defining type. You are creating new request - so name the method accordingly

File: basis/common-datamapper/src/main/java/com/softgroup/common/datamapper/JacksonDataMapper.java [hunk 8]
```diff
@@ -53,6 +53,15 @@ public ObjectMapper getMapper() {
 	}
 
 	@Override
+	public <T> T convertValue(Object object, Class<T> dataType) {
```
Comment (Dominys) [inline basis/common-datamapper/src/main/java/com/softgroup/common/datamapper/JacksonDataMapper.java]: @aptasinskij you added new method. Where is tests for this method ??

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java (hunk 6 above)
Comment (aptasinskij) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: You mean I need to pass the mapper as second argument in the constructor and mark constructor with Autowired annotation?

Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij Yes. If use constructor - pass injected values through constructor.

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java [hunk 9]
```diff
@@ -1,20 +1,39 @@
...
-	@Override
-	public String getName() {
-		return null;
-	}
+public abstract class AbstractRequestHandler<T extends RequestData, R extends ResponseData> implements RequestHandler {
+
+    private final Class<T> type;
+    private DataMapper mapper;
+
+    @Autowired
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRequestHandler.java]: @aptasinskij This is abstract class. Let Subclasses to decide autowire this or not.

File: basis/common-router/src/main/java/com/softgroup/common/router/api/factory/AbstractHandlerFactory.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Created by andrew on 27.02.17.
+ */
+public abstract class AbstractHandlerFactory<T extends Handler> implements HandlerFactory<T> {
+
+    private Map<String, T> handlersMap;
+
+    @Autowired
+    private List<T> handlersList;
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/factory/AbstractHandlerFactory.java]: @aptasinskij Lets use some ONE style. Field injection or constructor injection - on your taste.

Comment (aptasinskij) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/factory/AbstractHandlerFactory.java]: I would definitely choose constructor injection.

Comment (aptasinskij) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/factory/AbstractHandlerFactory.java]: Should I leave the decision about autowiring to subclasses?

Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/factory/AbstractHandlerFactory.java]: @aptasinskij yes.

File: basis/common-router/src/main/java/com/softgroup/common/router/api/factory/AbstractHandlerFactory.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+    private List<T> handlersList;
+
+    @Autowired
+    public AbstractHandlerFactory(List<T> handlersList){
+        this.handlersList = handlersList;
+    }
+
+    @PostConstruct
+    public void init(){
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/factory/AbstractHandlerFactory.java]: @aptasinskij Why you need post construct if you have constructor?

File: basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRouterHandler.java [hunk 12]
```diff
@@ -2,17 +2,17 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
 
 public abstract class AbstractRouterHandler<T extends Handler> implements RouterHandler {
 
-	@Override
-	public String getRouteKey(Request<?> msg) {
-		return null;
-	}
+    @Autowired
+    HandlerFactory<T> handlerFactory;
```
Comment (Dominys) [inline basis/common-router/src/main/java/com/softgroup/common/router/api/AbstractRouterHandler.java]: @aptasinskij You told me that you prefer constructor. But still ...

**Pull request merged**